File: backend/mt5_bridge.py
#!/usr/bin/env python3
import json
import sys
from datetime import datetime, timezone

import MetaTrader5 as mt5


TIMEFRAME_MAP = {
    "1m": mt5.TIMEFRAME_M1,
    "5m": mt5.TIMEFRAME_M5,
    "15m": mt5.TIMEFRAME_M15,
    "30m": mt5.TIMEFRAME_M30,
    "1h": mt5.TIMEFRAME_H1,
    "4h": mt5.TIMEFRAME_H4,
    "1d": mt5.TIMEFRAME_D1,
}


def fail(message: str, code: int = 1) -> None:
    print(json.dumps({"ok": False, "error": message}))
    sys.exit(code)


def emit(payload: dict) -> None:
    print(json.dumps({"ok": True, **payload}))


def ensure_initialized() -> None:
    if not mt5.initialize():
        fail(f"MT5 initialize failed: {mt5.last_error()}")


def ensure_symbol(symbol: str) -> None:
    info = mt5.symbol_info(symbol)
    if info is None:
        fail(f"Symbol not found: {symbol}")
    if not info.visible and not mt5.symbol_select(symbol, True):
        fail(f"Symbol not visible and could not be selected: {symbol}")


def get_account_info() -> dict:
    ensure_initialized()
    account = mt5.account_info()
    terminal = mt5.terminal_info()
    if account is None:
        raise RuntimeError("No MT5 account is currently connected")

    return {
        "account": {
            "login": account.login,
            "server": account.server,
            "broker": account.company,
            "name": account.name,
            "balance": account.balance,
            "equity": account.equity,
            "currency": account.currency,
            "leverage": account.leverage,
        },
        "terminal": {
            "name": getattr(terminal, "name", None),
            "company": getattr(terminal, "company", None),
            "connected": getattr(terminal, "connected", None),
            "trade_allowed": getattr(terminal, "trade_allowed", None),
            "path": getattr(terminal, "path", None),
        },
    }


def get_tick(symbol: str) -> dict:
    ensure_initialized()
    ensure_symbol(symbol)
    tick = mt5.symbol_info_tick(symbol)
    info = mt5.symbol_info(symbol)
    account = mt5.account_info()
    if tick is None or info is None:
        raise RuntimeError(f"No tick data for symbol: {symbol}")

    price = None
    if tick.bid and tick.ask:
        price = (tick.bid + tick.ask) / 2
    elif tick.last:
        price = tick.last

    if price is None:
        raise RuntimeError(f"No usable price for symbol: {symbol}")

    return {
        "symbol": symbol,
        "price": price,
        "bid": tick.bid,
        "ask": tick.ask,
        "last": tick.last,
        "spread": info.spread,
        "digits": info.digits,
        "time": tick.time,
        "time_iso": datetime.fromtimestamp(tick.time, tz=timezone.utc).isoformat(),
        "source": f"MT5-local:{account.server if account else 'unknown'}",
    }


def get_candles(symbol: str, interval: str, limit: int) -> dict:
    ensure_initialized()
    ensure_symbol(symbol)
    timeframe = TIMEFRAME_MAP.get(interval)
    if timeframe is None:
        raise ValueError(f"Unsupported timeframe: {interval}")

    rates = mt5.copy_rates_from_pos(symbol, timeframe, 0, limit)
    account = mt5.account_info()
    if rates is None:
        raise RuntimeError(f"No rates returned for {symbol} {interval}")

    candles = []
    for rate in rates:
        candles.append(
            {
                "time": int(rate["time"]),
                "open": float(rate["open"]),
                "high": float(rate["high"]),
                "low": float(rate["low"]),
                "close": float(rate["close"]),
                "tick_volume": int(rate["tick_volume"]),
            }
        )

    return {
        "symbol": symbol,
        "interval": interval,
        "count": len(candles),
        "candles": candles,
        "source": f"MT5-local:{account.server if account else 'unknown'}",
    }


def main() -> None:
    if len(sys.argv) < 2:
        fail("Usage: mt5_bridge.py <account-info|tick|candles> [args...]")

    command = sys.argv[1]
    try:
        if command == "account-info":
            emit(get_account_info())
        elif command == "tick":
            if len(sys.argv) < 3:
                fail("Usage: mt5_bridge.py tick <SYMBOL>")
            emit(get_tick(sys.argv[2].upper()))
        elif command == "candles":
            if len(sys.argv) < 5:
                fail("Usage: mt5_bridge.py candles <SYMBOL> <INTERVAL> <LIMIT>")
            emit(get_candles(sys.argv[2].upper(), sys.argv[3], int(sys.argv[4])))
        else:
            fail(f"Unknown command: {command}")
    finally:
        mt5.shutdown()


if __name__ == "__main__":
    main()

File: app/lib/mt5Bridge.ts
import { execFile } from 'node:child_process';
import path from 'node:path';

export interface Mt5AccountInfo {
  account: {
    login: number;
    server: string;
    broker: string;
    name: string;
    balance: number;
    equity: number;
    currency: string;
    leverage: number;
  };
  terminal: {
    name?: string | null;
    company?: string | null;
    connected?: boolean | null;
    trade_allowed?: boolean | null;
    path?: string | null;
  };
}

export interface Mt5Tick {
  symbol: string;
  price: number;
  bid: number;
  ask: number;
  last: number;
  spread: number;
  digits: number;
  time: number;
  time_iso: string;
  source: string;
}

export interface Mt5Candle {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  tick_volume: number;
}

export interface Mt5CandleBatch {
  symbol: string;
  interval: string;
  count: number;
  candles: Mt5Candle[];
  source: string;
}

type BridgeResponse<T> =
  | ({ ok: true } & T)
  | { ok: false; error: string };

const bridgeScript = path.join(process.cwd(), 'backend', 'mt5_bridge.py');
const pythonExecutable = process.env.MT5_BRIDGE_PYTHON || 'python';
const bridgeUrl = process.env.MT5_BRIDGE_URL?.replace(/\/$/, '');
const bridgeToken = process.env.MT5_BRIDGE_TOKEN;

function getRemoteBridgeUrl(): string | null {
  if (!bridgeUrl || !bridgeToken) return null;

  try {
    const url = new URL(bridgeUrl);
    const isLocalHttp = url.protocol === 'http:' && ['localhost', '127.0.0.1'].includes(url.hostname);

    // Vercel must never send an MT5 bridge token over an unencrypted connection.
    if (process.env.VERCEL === '1' && url.protocol !== 'https:') return null;
    if (url.protocol !== 'https:' && !isLocalHttp) return null;

    return url.toString().replace(/\/$/, '');
  } catch {
    return null;
  }
}

function execFileAsync(file: string, args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    execFile(file, args, { timeout: 15000, maxBuffer: 1024 * 1024 }, (error, stdout, stderr) => {
      if (error) {
        reject(new Error(stderr?.trim() || error.message));
        return;
      }
      resolve(stdout);
    });
  });
}

async function callMt5Bridge<T>(args: string[]): Promise<T | null> {
  const remoteUrl = getRemoteBridgeUrl();

  if (remoteUrl) {
    try {
      const remoteArgs = args[0] === 'account-info' ? ['account'] : args;
      const response = await fetch(`${remoteUrl}/${remoteArgs.map(encodeURIComponent).join('/')}`, {
        headers: { Authorization: `Bearer ${bridgeToken}` },
        cache: 'no-store',
        signal: AbortSignal.timeout(10_000),
      });

      if (!response.ok) return null;
      const parsed = await response.json() as BridgeResponse<T>;
      return parsed.ok ? parsed as T : null;
    } catch {
      return null;
    }
  }

  // Serverless hosts cannot access the Windows MT5 terminal or execute MetaTrader5.
  if (process.env.VERCEL === '1') return null;

  try {
    const stdout = await execFileAsync(pythonExecutable, [bridgeScript, ...args]);
    const parsed = JSON.parse(stdout.trim()) as BridgeResponse<T>;
    if (!parsed.ok) return null;

    return parsed as T;
  } catch {
    return null;
  }
}

export async function getMt5AccountInfo(): Promise<Mt5AccountInfo | null> {
  return callMt5Bridge<Mt5AccountInfo>(['account-info']);
}

export async function getMt5Tick(symbol: string): Promise<Mt5Tick | null> {
  return callMt5Bridge<Mt5Tick>(['tick', symbol.toUpperCase()]);
}

export async function getMt5Candles(symbol: string, interval: string, limit: number): Promise<Mt5CandleBatch | null> {
  return callMt5Bridge<Mt5CandleBatch>(['candles', symbol.toUpperCase(), interval, String(limit)]);
}

export function isMt5BridgeConfigured(): boolean {
  return getRemoteBridgeUrl() !== null;
}
